### ISSUE TITLE
error sending email with Mailgun

### ISSUE THREAD

**Issue by dwrensha** [sandstorm-io/sandstorm#1148]
I'm trying to send email from a RoundCube grain on a Sandstorm server using Mailgun for outgoing email. I'm pretty sure all of this worked without a problem as of a month or two ago (and it still seems to work fine on Oasis). Now when I try to send, RoundCube pops up a modal saying "An error occurred!" and I see the following in my sandstorm.log:

```
Error sending e-mail: DeliveryError: Message delivery failed: 500 Error: bad syntax
    at Object.Future.wait (/programs/server/node_modules/fibers/future.js:398:15)
    at smtpSend (packages/sandstorm-email/email.js:101:1)
    at Object.SandstormEmail.rawSend (packages/sandstorm-email/email.js:176:1)
    at app/server/drivers/mail.js:282:20
    at app/server/proxy.js:111:17
    at app/server/proxy.js:103:3
    at runWithEnvironment (packages/meteor/dynamics_nodejs.js:108:1)
    - - - - -
500 Command not implemented
    at SMTPConnectionPool._onConnectionReady (/programs/server/npm/sandstorm-email/node_modules/simplesmtp/lib/pool.js:287:21)
    at SMTPClient.emit (events.js:98:17)
    at SMTPClient._actionStream (/programs/server/npm/sandstorm-email/node_modules/simplesmtp/lib/client.js:946:14)
    at SMTPClient._onData (/programs/server/npm/sandstorm-email/node_modules/simplesmtp/lib/client.js:329:29)
    at CleartextStream.emit (events.js:95:17)
    at CleartextStream.<anonymous> (_stream_readable.js:765:14)
    at CleartextStream.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:427:10)
    at _stream_readable.js:420:7
    at process._tickCallback (node.js:442:13)
```

Oddly, the message does actually reach the intended recipient. So I end up in a bad state when my RoundCube client thinks that the send failed, but it really succeeded.

The MailComposer object, as passed to `smtpSend()`, looks like:

```
{
    "_alternatives": [],
    "_attachments": [],
    "_cacheOutput": false,
    "_currentStream": false,
    "_dkim": false,
    "_envelope": {
        "from": [
            "0j1lxpsvh1bk8ayq2lff@sandstorm-lithium.dwrensha.ws"
        ],
        "stamp": [
            "",
            ""
        ],
        "to": [
            "david@sandstorm.io"
        ],
        "userDefined": true
    },
    "_events": {},
    "_gencounter": 0,
    "_headers": {
        "Date": "Tue, 10 Nov 2015 16:04:45 GMT",
        "Message-Id": "<d6d921f21cc41247d1890363c472f35f@63d551d2944f4cc2eb9695209a3a17f5.sandstorm-lithium.dwrensha.ws>"
    },
    "_lineEnding": "\r\n",
    "_maxListeners": 10,
    "_message": {
        "bcc": "",
        "body": "hello?\r\n.\r\n",
        "cc": "",
        "envelope": {
            "from": "0j1lxpsvh1bk8ayq2lff@sandstorm-lithium.dwrensha.ws",
            "stamp": "Postage paid, Par Avion",
            "to": [
                "david@sandstorm.io"
            ]
        },
        "from": "\"Alice Dev Admin\" <0j1lxpsvh1bk8ayq2lff@sandstorm-lithium.dwrensha.ws>",
        "replyTo": "",
        "subject": "testing",
        "to": "david@sandstorm.io"
    },
    "_outputBuffer": "",
    "_relatedAttachments": [],
    "domain": null,
    "options": {
        "charset": "utf-8",
        "identityString": "mailcomposer"
    },
    "readable": true
}
```

**Comment by dwrensha** [comment]
On further reflection, I am less certain that sending from Roundcube has ever worked correctly for me on this server. Email login definitely works fine, however.

**Comment by dwrensha** [comment]
It appears that somehow this sequence in `mc._message.body` is what's causing the trouble: "\r\n.\r\n"

If I replace that field with "does this work?\r\nI hope so.\r\n\r\n" right before sending, I get no error.

**Comment by dwrensha** [comment]
Values for the `body` field that succeed: `["", "a", "..", "\r\n", "a.\r\n", "a.\r\n\r\n", "..\r\n"]`
Values for the `body` field that fail: `[".", "\r\n." ".\r\n", ".\r\n\r\n"]`

It's looking to me like the problem is lines that consist only of "."

**Comment by geofft** [comment]
Your SMTP client is supposed to escape those by adding an extra dot to the beginning of lines that start with a dot. (So you're presumably getting a 500 from whatever remainder of the body gets interpreted as a command.) See [RFC 5321 section 4.5.2](http://tools.ietf.org/html/rfc5321#section-4.5.2).

A casual look at simplesmtp doesn't find anything that implements this logic, but it does find a mention that simplesmtp is deprecated in favor of [smtp-connection](https://www.npmjs.com/package/smtp-connection), which [does](https://github.com/andris9/smtp-connection/blob/master/src/data-stream.js#L27).

**Comment by dwrensha** [comment]
Fixed by #1149. See also https://github.com/jparyani/roundcubemail/issues/4.
